dateparser/date.py

```diff
@@ -87,55 +90,145 @@ def get_last_day_of_month(year, month):
-    data = {'period': 'day', 'date_obj': None}
-
     if isinstance(date_string, unicode):
         date_string = date_string.encode('utf-8')
+
+    period = 'day'
     for date_format in date_formats:
         try:
-            try:
-                date_obj = datetime.strptime(date_string, date_format)
-
-                # If format does not include the day, use last day of the month
-                # instead of first, because the first is usually out of range.
-                if '%d' not in date_format:
-                    data['period'] = 'month'
-                    date_obj = date_obj.replace(
-                        day=get_last_day_of_month(date_obj.year, date_obj.month))
-
-                if not ('%y' in date_format or '%Y' in date_format):
-                    today = datetime.today()
-                    date_obj = date_obj.replace(year=today.year)
-
-            except ValueError:
-                alt_parser = alt_parser if alt_parser else DateParser()
-                date_obj = alt_parser.parse(date_string, date_format=date_format)
-            if final_call:
-                data['date_obj'] = date_obj
-                return data
-            else:
-                return date_obj
+            date_obj = datetime.strptime(date_string, date_format)
         except ValueError:
             continue
+        else:
+            # If format does not include the day, use last day of the month
+            # instead of first, because the first is usually out of range.
+            if '%d' not in date_format:
+                period = 'month'
+                date_obj = date_obj.replace(
+                    day=get_last_day_of_month(date_obj.year, date_obj.month))
+
+            if not ('%y' in date_format or '%Y' in date_format):
+                today = datetime.today()
+                date_obj = date_obj.replace(year=today.year)
+
+            return {'date_obj': date_obj, 'period': period}
     else:
-        if final_call:
-            return data
+        return {'date_obj': None, 'period': period}
+
+
+class _DateLanguageParser(object):
+    def __init__(self, language, date_string, date_formats):
+        self.language = language
+        self.date_string = date_string
+        self.date_formats = date_formats
+        self._translated_date = None
+        self._translated_date_with_formatting = None
+
+    @classmethod
+    def parse(cls, language, date_string, date_formats=None):
+        instance = cls(language, date_string, date_formats)
+        return instance._parse()
+
+    def _parse(self):
+        for parser in (
+            self._try_timestamp,
+            self._try_freshness_parser,
+            self._try_given_formats,
+            self._try_dateutil_parser,
+            self._try_hardcoded_formats,
+        ):
+            date_obj = parser()
+            if self._is_valid_date_obj(date_obj):
+                return date_obj
+        else:
+            return None
```

Comment:
Let me explain why i do like for-else clause. It is explicitly stating "Hey, we expected something to happen in the loop, but as it did not happen, here is our plan 'B'". While it does the same job, it is a great way to organize code, because without the else, there are no expectations set.